### PR TITLE
fix(plugin-catch-links): handle pathPrefix

### LINKED ISSUE #8907: gatsby-plugin-catch-links fails if pathPrefix exists
## Description

If pathPrefix exists, *gatsby-plugin-catch-links* works incorrectly for links in Markdown files. As a result, the prefix has been added twice, once by *gatsby-transformer-remark*, and once by `navigate`.

### Steps to reproduce

Demo project: https://github.com/gera2ld/demo-gatsby

```sh
# build with prefix `/public`
$ yarn build --prefix-paths

# start a local server
$ python3 -m http.server

# open http://localhost:8000/public/
```

### Expected result

Click on `page-2`, the expected behavior is to navigate to `/public/page-2/`.

### Actual result

Navigated to `/public/public/page-2/` and got 404.

### Environment

```
  System:
    OS: macOS High Sierra 10.13.6
    CPU: x64 Intel(R) Core(TM) i7-4770HQ CPU @ 2.20GHz
    Shell: 5.3 - /bin/zsh
  Binaries:
    Node: 10.3.0 - ~/.nvm/versions/node/v10.3.0/bin/node
    Yarn: 1.10.1 - /usr/local/bin/yarn
    npm: 6.1.0 - ~/.nvm/versions/node/v10.3.0/bin/npm
  Browsers:
    Chrome: 69.0.3497.100
    Firefox: 63.0
    Safari: 12.0
  npmPackages:
    gatsby: ^2.0.0 => 2.0.19
    gatsby-plugin-catch-links: ^2.0.4 => 2.0.4
    gatsby-plugin-manifest: ^2.0.2 => 2.0.4
    gatsby-plugin-offline: ^2.0.5 => 2.0.5
    gatsby-plugin-react-helmet: ^3.0.0 => 3.0.0
    gatsby-source-filesystem: ^2.0.2 => 2.0.2
    gatsby-transformer-remark: ^2.1.6 => 2.1.6
```

### PR DESCRIPTION
**Pull request by ahinrichs**
Remove pathPrefix before calling navigate, closes #8907

### CODE REVIEW

**Comment by pieh** [comment]
Should we add some tests for pathPrefix scenarios?

**Comment by JimTheMan** [comment]
hey, I have experience writing tests, and also I need this PR to be merged so my site will work. lol

...but I have zero familiarity with the gatsby codebase and not sure exactly what it is that needs to be tested and where to put the tests.

**Comment by JimTheMan** [comment]
Also [this](https://github.com/gatsbyjs/gatsby/blob/0e2b88a140003f0825f85d4bd5c42ecfa041b655/packages/gatsby-plugin-catch-links/src/__tests__/catch-links.js#L44). Why are we passing "jest.fn()" to the method `routeThroughBrowserOrApp`? Can we have more cases to run through routeThroughBrowserOrApp?

**Comment by ahinrichs** [comment]
Actually I can't think of any situation where this PR makes things worse. I'm just reusing the regex that was created inside `pathIsNotHandledByApp`. Therefore I create it in the callling function and passes it as a parameter to `pathIsNotHandledByApp`. The only semantic change is to replace `/<pathPrefix>/` with `/` before calling `hrefHandler` if the catched link starts with it. So that `navigate` will get the unprefixed path as it expects. In builds without pathPrefix this is `/` to `/`. And in builds with pathPrefix things don't work at all ATM.

I wouldn't hold back the merge to wait for having tests. But it is a regression from #8289. And maybe tests would have helped to avoid this. Maybe merge and create an issue that this plugin needs more tests?

**Comment by ahinrichs** [comment]
Found what I wanted in gatsby-link tests

**Comment by JimTheMan** [comment]
nice @ahinrichs. I like your idea for the test, but how can you know from the test when window.history recieves an event?

**Comment by ahinrichs** [comment]
For the best test one really should somehow mock `window.history`. This test is what `gatsby-link.navigate` - which is what is used by this module - finally passes to `@reach/router.navigate`. So it should be good enough to cover changes like #8289.

**Comment by ahinrichs** [comment]
This was just a try. The integration test worked here but failed in circle.ci. So I reverted.

**Comment by ahinrichs** [comment]
BTW, I always knew I will fall in love with writing tests the moment I start doing it :-)

File: packages/gatsby-plugin-catch-links/src/__tests__/catch-links.js [hunk 1]
```diff
@@ -351,3 +358,82 @@ describe(`navigation is routed through gatsby if the destination href`, () => {
...
+    const clickEvent = new MouseEvent(`click`, {
+      bubbles: true,
+      cancelable: true,
+      view: window,
+    })
+
+    hrefHandler.mockImplementation((path) => {
+      getNavigate()(path)
+      expect(global.___navigate).toHaveBeenCalledWith(
+        `${pathPrefix}/someSubPath`,
```
Comment (pieh) [inline packages/gatsby-plugin-catch-links/src/__tests__/catch-links.js]: will this test actually fail if you revert changes to `src/catch-links.js`? gotta say it's kind of confusing one to read ;)

Comment (ahinrichs) [reply packages/gatsby-plugin-catch-links/src/__tests__/catch-links.js]: Just the one of them that simulates a build with --prefix-paths. That's the current state: module works in default builds but not in builds active prefix. But I wanted to catch the case too if any change would break the non prefix processing. As I said, these two are more of an integration test. I just don't know why I can't `import "gatsby-link"` there.

And yes, it was confusing. I tend to leave otherones code untouched. And this was a straight combination of the existing tests and one of the gatsby-link tests. I did some refactor and added comments.

**Comment by ahinrichs** [comment]
I created another PR #9051. With this circleci setup the test from commit 9e09c65 succeeds as an integration test (see #9050). Acutally we could still merge this PR in the actual state to get the plugin fixed and maybe--if I'm right with the circle ci setup--could change this PR tests the to integration tests in another step.

**Comment by chasemccoy** [comment]
@ahinrichs @pieh any ETA on getting this merged? This bug is causing some big headaches for me.

**Comment by pieh** [comment]
@chasemccoy I will be testing this locally today (mostly because those tests are kinda confusing to me :P)

**Comment by ahinrichs** [comment]
Just on last change to `import { navigate } from "gatsby"` instead of `from "gatsby-link"`.

I'm pretty confident, that this PR first and most important fixes the issue and second the tests at least do no harm. So I think it could be merged.

The only thing--beside the new tests being a bit abstract--that is not ideal is the import from `gatsby-link` resp. `gatsby`. That way it isn't a pure unit test anymore. But that could be done in the next step.

@pieh If you feel uncomfortable with the tests, would it help if we just merge the change in the regular code?

**Comment by pieh** [comment]
> @pieh If you feel uncomfortable with the tests, would it help if we just merge the change in the regular code?

No, we definitely want some kind of tests to catch potential regressions, I just need to dig a bit and see if it actually tests stuff ;)

**Comment by pieh** [comment]
So it tests correctly, little verbose but better to have this in (+fix obviously) and try to figure out nicer tests (or some refactoring to not repeat a lot of code there).

I just formatted code and skipped empty tests (we shouldn't report success on empty tests - this just gives false sense of coverage)

**Comment by chasemccoy** [comment]
Woohoo! Thank y'all for pushing this through, this will save me a lot of headaches.

**Comment by ahinrichs** [comment]
Actually - as an test beginner - I followed what was there. I first had a version where the test itself was in a function. But I assumed the code repetition in the tests before was for a reason. So I rewrote them to be two independent code blocks. 

In the end I discovered the e2e test setup and I think this is best placed there. But this will take some days.

**Comment by ahinrichs** [comment]
The empty tests where to give the output some kind of "story telling" impression. They were already there. But skip doesn't make much sense to me. I would then prefer to remove them (and maybe rephrase some of the other texts to stand on their own).
